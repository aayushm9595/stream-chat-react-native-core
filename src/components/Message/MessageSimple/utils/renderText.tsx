import React from 'react';
// @ts-expect-error
import Markdown from '@stream-io/react-native-simple-markdown';
import anchorme from 'anchorme';
import truncate from 'lodash/truncate';

import type { MarkdownStyle } from '../../../../styles/themeConstants';
import type { Message } from '../../../MessageList/utils/insertDates';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../../types/types';

const defaultMarkdownStyles = {
  inlineCode: {
    backgroundColor: '#F3F3F3',
    borderColor: '#dddddd',
    color: 'red',
    fontSize: 13,
    padding: 3,
    paddingHorizontal: 5,
  },
  link: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
  url: {
    color: 'blue',
    textDecorationLine: 'underline',
  },
};

export type RenderTextParams<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  markdownRules: UnknownType;
  markdownStyles: MarkdownStyle;
  message: Message<At, Ch, Co, Ev, Me, Re, Us>;
};

export const renderText = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  params: RenderTextParams<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { markdownRules, markdownStyles, message } = params;

  // take the @ mentions and turn them into markdown?
  // translate links
  const { mentioned_users = [], text } = message;

  if (!text) return null;

  let newText = text.trim();
  const urls = anchorme(newText, {
    list: true,
  });

  for (const urlInfo of urls) {
    const displayLink = truncate(urlInfo.encoded.replace(/^(www\.)/, ''), {
      length: 200,
      omission: '...',
    });
    const markdown = `[${displayLink}](${urlInfo.protocol}${urlInfo.encoded})`;
    newText = newText.replace(urlInfo.raw, markdown);
  }

  if (mentioned_users.length) {
    for (let i = 0; i < mentioned_users.length; i++) {
      const username = mentioned_users[i].name || mentioned_users[i].id;
      const markdown = `**@${username}**`;
      const regEx = new RegExp(`@${username}`, 'g');
      newText = newText.replace(regEx, markdown);
    }
  }

  newText = newText.replace(/[<&"'>]/g, '\\$&');
  const styles = {
    ...defaultMarkdownStyles,
    ...markdownStyles,
  };

  return (
    <Markdown rules={markdownRules} styles={styles}>
      {newText}
    </Markdown>
  );
};
