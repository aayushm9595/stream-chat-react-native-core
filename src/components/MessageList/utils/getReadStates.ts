import {
  InsertDatesResponse,
  isDateSeparator,
  Message,
} from '../utils/insertDates';

import type { ImmutableDate } from 'seamless-immutable';
import type { UserResponse } from 'stream-chat';

import type { ChannelContextValue } from '../../../contexts/channelContext/ChannelContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const getReadStates = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  messages: InsertDatesResponse<At, Ch, Co, Ev, Me, Re, Us>,
  read?: ChannelContextValue<At, Ch, Co, Ev, Me, Re, Us>['read'],
) => {
  const readData = messages.reduce((acc, cur) => {
    if (!isDateSeparator(cur) && cur.id) {
      acc[cur.id] = [];
    }
    return acc;
  }, {} as { [key: string]: UserResponse<Us>[] });

  const filteredMessagesReversed = messages
    .filter((msg) => !isDateSeparator(msg) && msg.updated_at)
    .reverse() as Array<
    Message<At, Ch, Co, Ev, Me, Re, Us> & {
      updated_at: string | ImmutableDate;
    }
  >;

  if (read) {
    /**
     * Channel read state is stored by user
     */
    for (const readState of Object.values(read)) {
      /**
       * If no last read continue
       */
      if (!readState.last_read) {
        continue;
      }

      /**
       * Array is in reverse order so newest message is at 0,
       * we find the index of the first message that is older
       * than the last read and then set last read to that, or
       * if there are no newer messages, the first message is
       * last read message.
       */
      const userLastReadMsgId = filteredMessagesReversed.find(
        (msg) => msg.updated_at < readState.last_read,
      )?.id;

      /**
       * If there there is a last read message add the user
       * to the array of last reads for that message
       */
      if (userLastReadMsgId) {
        readData[userLastReadMsgId] = [
          ...readData[userLastReadMsgId],
          readState.user as UserResponse<Us>,
        ];
      }
    }
  }

  return readData;
};
