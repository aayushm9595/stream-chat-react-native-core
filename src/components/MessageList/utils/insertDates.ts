import type { ChannelState } from 'stream-chat';

import type {
  MessagesContextValue,
  MessageWithDates,
} from '../../../contexts/messagesContext/MessagesContext';
import type { ThreadContextValue } from '../../../contexts/threadContext/ThreadContext';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export type DateSeparator<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  date: Message<At, Ch, Co, Ev, Me, Re, Us>['created_at'] | string;
  type: 'message.date';
};

export type MessageOrDate<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> =
  | Message<At, Ch, Co, Ev, Me, Re, Us>
  | DateSeparator<At, Ch, Co, Ev, Me, Re, Us>;

export type InsertDatesResponse<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = MessageOrDate<At, Ch, Co, Ev, Me, Re, Us>[];

export const isDateSeparator = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  message: MessageOrDate<At, Ch, Co, Ev, Me, Re, Us>,
): message is DateSeparator<At, Ch, Co, Ev, Me, Re, Us> =>
  message.type === 'message.date';

export type Message<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> =
  | ReturnType<ChannelState<At, Ch, Co, Ev, Me, Re, Us>['messageToImmutable']>
  | MessageWithDates<At, Ch, Co, Me, Re, Us>;

export const insertDates = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  messages:
    | MessagesContextValue<At, Ch, Co, Ev, Me, Re, Us>['messages']
    | ThreadContextValue<At, Ch, Co, Ev, Me, Re, Us>['threadMessages'],
) => {
  const newMessages: InsertDatesResponse<At, Ch, Co, Ev, Me, Re, Us> = [];
  if (messages.length === 0) {
    return newMessages;
  }

  for (const [i, message] of messages.entries()) {
    /**
     * If message read or deleted don't consider for date labels
     */
    if (message.type === 'message.read' || message.deleted_at) {
      newMessages.push(message);
      continue;
    }

    /**
     * Get the date of the current message and create
     * variable for previous date (day)
     */
    const messageDate = message.created_at.getDay();

    /**
     * If this is not the last entry in the messages array
     * set the previous message date (day) to the date of the next
     * message in the array
     */
    const prevMessageDate =
      i < messages.length - 1
        ? messages[i + 1].created_at.getDay()
        : messageDate;

    /**
     * Before the first message insert a date object
     */
    if (i === 0) {
      newMessages.push(
        {
          date: message.created_at,
          type: 'message.date',
        } as DateSeparator<At, Ch, Co, Ev, Me, Re, Us>,
        message,
      );

      /**
       * If the date (day) has changed between two messages
       * insert a date object
       */
    } else if (messageDate !== prevMessageDate) {
      newMessages.push(message, {
        date: messages[i + 1].created_at,
        type: 'message.date',
      } as DateSeparator<At, Ch, Co, Ev, Me, Re, Us>);

      /**
       * Otherwise just add the message
       */
    } else {
      newMessages.push(message);
    }
  }

  return newMessages;
};
