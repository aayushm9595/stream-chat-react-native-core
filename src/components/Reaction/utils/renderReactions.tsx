import React from 'react';
import { Text } from 'react-native';

import type { ReactionResponse } from 'stream-chat';

import type { LatestReactions, Reaction } from '../ReactionList';
import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../../types/types';

export const renderReactions = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  reactions: LatestReactions<At, Ch, Co, Me, Re, Us>,
  supportedReactions: Reaction[],
) => {
  const reactionsByType: {
    [key: string]: ReactionResponse<Re, Us>[];
  } = {};
  reactions?.map((reaction) => {
    if (!reactionsByType[reaction.type]) {
      return (reactionsByType[reaction.type] = [reaction]);
    }
    return (reactionsByType[reaction.type] = [
      ...reactionsByType[reaction.type],
      reaction,
    ]);
  });

  const emojiDataByType: { [key: string]: Reaction } = {};
  const reactionTypes = supportedReactions.map((supportedReaction) => {
    emojiDataByType[supportedReaction.id] = supportedReaction;
    return supportedReaction.id;
  });

  return Object.keys(reactionsByType).map((type) =>
    reactionTypes.indexOf(type) > -1 ? (
      <Text key={type} testID={type}>
        {emojiDataByType[type].icon}
      </Text>
    ) : null,
  );
};
