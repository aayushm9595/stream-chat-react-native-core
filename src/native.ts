import type { NetInfoSubscription } from '@react-native-community/netinfo';

const fail = () => {
  throw Error(
    'Native handler was not registered, you should import stream-chat-expo or stream-chat-react-native',
  );
};

type NetInfo = {
  addEventListener: (
    listener: (isConnected: boolean) => void,
  ) => NetInfoSubscription | never;
  fetch: (requestedInterface?: string | undefined) => Promise<boolean> | never;
};

export let NetInfo: NetInfo = {
  addEventListener: fail,
  fetch: fail,
};

type PickImage = ({
  compressImageQuality,
  maxNumberOfFiles,
}: {
  compressImageQuality?: number;
  maxNumberOfFiles?: number;
}) => Promise<{ cancelled: boolean; images?: { uri: string }[] }> | never;
export let pickImage: PickImage = fail;

type PickDocument = ({
  maxNumberOfFiles,
}: {
  maxNumberOfFiles?: number;
}) =>
  | Promise<{
      cancelled: boolean;
      docs?: {
        name: string;
        size?: number | string;
        type?: string;
        uri?: string;
      }[];
    }>
  | never;
export let pickDocument: PickDocument = fail;

type Handlers = {
  NetInfo?: NetInfo;
  pickDocument?: PickDocument;
  pickImage?: PickImage;
};
export const registerNativeHandlers = (handlers: Handlers) => {
  if (handlers.NetInfo) {
    NetInfo = handlers.NetInfo;
  }

  if (handlers.pickImage) {
    pickImage = handlers.pickImage;
  }

  if (handlers.pickDocument) {
    pickDocument = handlers.pickDocument;
  }
};
