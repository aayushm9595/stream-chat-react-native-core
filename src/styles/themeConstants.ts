import type { ImageStyle, StyleProp, TextStyle, ViewStyle } from 'react-native';
import type { CSSObject } from 'styled-components';

export const BASE_FONT_SIZE = 16;

export const Colors = {
  danger: '#EDD8DD',
  light: '#EBEBEB',
  primary: '#006cff',
  secondary: '#111',
  textDark: 'rgba(0,0,0,1)',
  textGrey: 'rgba(0,0,0,0.5)',
  textLight: 'white',
  transparent: 'transparent',
};

export type CSS = {
  css?: CSSObject;
};

export type MarkdownStyle = Partial<{
  blockQuoteBar: StyleProp<ViewStyle>;
  blockQuoteSection: StyleProp<ViewStyle>;
  blockQuoteSectionBar: StyleProp<ViewStyle>;
  blockQuoteText: StyleProp<ViewStyle>;
  codeBlock: StyleProp<TextStyle>;
  del: StyleProp<TextStyle>;
  em: StyleProp<TextStyle>;
  heading: StyleProp<TextStyle>;
  heading1: StyleProp<TextStyle>;
  heading2: StyleProp<TextStyle>;
  heading3: StyleProp<TextStyle>;
  heading4: StyleProp<TextStyle>;
  heading5: StyleProp<TextStyle>;
  heading6: StyleProp<TextStyle>;
  hr: StyleProp<ViewStyle>;
  image: StyleProp<ImageStyle>;
  inlineCode: StyleProp<TextStyle>;
  link: StyleProp<TextStyle>;
  listItem: StyleProp<ViewStyle>;
  listItemBulletType: string;
  listItemButton: StyleProp<TextStyle>;
  listItemNumber: StyleProp<TextStyle>;
  listItemText: StyleProp<TextStyle>;
  mailTo: StyleProp<TextStyle>;
  paragraph: StyleProp<ViewStyle>;
  strong: StyleProp<TextStyle>;
  table: StyleProp<ViewStyle>;
  tableHeader: StyleProp<ViewStyle>;
  tableHeaderCell: StyleProp<ViewStyle>;
  tableRow: StyleProp<ViewStyle>;
  tableRowCell: StyleProp<ViewStyle>;
  tableRowLast: StyleProp<ViewStyle>;
  text: StyleProp<TextStyle>;
  u: StyleProp<TextStyle>;
  video: StyleProp<ImageStyle>;
  view: StyleProp<ViewStyle>;
}>;

export type Theme = {
  avatar: {
    container: CSS;
    fallback: CSS;
    image: CSS;
    text: CSS;
  };
  channelListFooterLoadingIndicator: {
    container: CSS;
  };
  channelListHeaderErrorIndicator: {
    container: CSS;
    errorText: CSS;
  };
  channelListHeaderNetworkDownIndicator: {
    container: CSS;
    errorText: CSS;
  };
  channelPreview: {
    container: CSS;
    date: CSS;
    details: CSS;
    detailsTop: CSS;
    message: CSS & {
      color: string;
      fontWeight: string;
      unreadColor: string;
      unreadFontWeight: string;
    };
    title: CSS;
  };
  closeButton: {
    container: CSS;
  };
  colors: typeof Colors;
  iconBadge: {
    container: CSS;
    icon: CSS;
    iconInner: CSS;
    unreadCount: CSS;
  };
  iconSquare: {
    container: CSS;
  };
  loadingErrorIndicator: {
    container: CSS;
    errorText: CSS;
    retryText: CSS;
  };
  loadingIndicator: {
    container: CSS;
    loadingText: CSS;
  };
  message: {
    actions: {
      button: CSS & {
        defaultBackgroundColor: string;
        defaultBorderColor: string;
        primaryBackgroundColor: string;
        primaryBorderColor: string;
      };
      buttonText: CSS & {
        defaultColor: string;
        primaryColor: string;
      };
      container: CSS;
    };
    actionSheet: {
      buttonContainer: CSS;
      buttonText: CSS;
      cancelButtonContainer: CSS;
      cancelButtonText: CSS;
      titleContainer: CSS;
      titleText: CSS;
    };
    avatarWrapper: {
      container: CSS;
      spacer: CSS;
    };
    card: {
      container: CSS;
      cover: CSS;
      footer: CSS & {
        description: CSS;
        link: CSS;
        logo: CSS;
        title: CSS;
      };
    };
    container: CSS;
    content: {
      container: CSS & {
        borderRadiusL: number;
        borderRadiusS: number;
      };
      containerInner: CSS;
      deletedContainer: CSS;
      deletedText: CSS;
      errorContainer: CSS & {
        backgroundColor: string;
      };
      // Available options for styling text: https://github.com/CharlesMangwa/react-native-simple-markdown/tree/next#styles-1
      markdown: MarkdownStyle;
      metaContainer: CSS;
      metaText: CSS;
      textContainer: CSS & {
        borderRadiusL: number;
        borderRadiusS: number;
        leftBorderColor: string;
        leftBorderWidth: number;
        rightBorderColor: string;
        rightBorderWidth: number;
      };
    };
    file: {
      container: CSS;
      details: CSS;
      icon: CSS;
      size: CSS;
      title: CSS;
    };
    gallery: {
      doubleSize: number;
      galleryContainer: CSS;
      halfSize: number;
      header: {
        button: CSS;
        container: CSS;
      };
      imageContainer: CSS;
      single: CSS;
      size: number;
      width: number;
    };
    reactionList: {
      container: CSS;
      reactionCount: CSS;
    };
    reactionPicker: {
      column: CSS;
      container: CSS;
      containerView: CSS;
      emoji: CSS;
      reactionCount: CSS;
      text: CSS;
    };
    replies: {
      container: CSS;
      image: CSS;
      messageRepliesText: CSS;
    };
    status: {
      checkMark: CSS;
      deliveredCircle: CSS;
      deliveredContainer: CSS;
      readByContainer: CSS;
      readByCount: CSS;
      sendingContainer: CSS;
      sendingImage: CSS;
      spacer: CSS;
    };
  };
  messageInput: {
    actionSheet: {
      buttonContainer: CSS;
      buttonText: CSS;
      titleContainer: CSS;
      titleText: CSS;
    };
    attachButton: CSS;
    attachButtonIcon: CSS;
    container: CSS & {
      conditionalPadding: number;
    };
    editingBoxContainer: CSS;
    editingBoxHeader: CSS;
    editingBoxHeaderTitle: CSS;
    fileUploadPreview: {
      attachmentContainerView: CSS;
      attachmentView: CSS;
      container: CSS;
      dismiss: CSS;
      dismissImage: CSS;
      filenameText: CSS;
      itemContainer: CSS;
    };
    imageUploadPreview: {
      container: CSS;
      dismiss: CSS;
      dismissImage: CSS;
      itemContainer: CSS;
      upload: CSS;
    };
    inputBox: CSS;
    inputBoxContainer: CSS;
    sendButton: CSS;
    sendButtonIcon: CSS;
    suggestions: {
      command: {
        args: CSS;
        container: CSS;
        description: CSS;
        title: CSS;
        top: CSS;
      };
      container: CSS & {
        itemHeight: number;
        maxHeight: number;
      };
      item: CSS;
      mention: {
        container: CSS;
        name: CSS;
      };
      separator: CSS;
      title: CSS;
      wrapper: CSS;
    };
    uploadProgressIndicator: {
      container: CSS;
      overlay: CSS;
    };
  };
  messageList: {
    dateSeparator: {
      container: CSS;
      date: CSS;
      dateText: CSS;
      line: CSS;
    };
    errorNotification: CSS;
    errorNotificationText: CSS;
    eventIndicator: {
      date: CSS;
      memberUpdateContainer: CSS;
      memberUpdateText: CSS;
      memberUpdateTextContainer: CSS;
    };
    listContainer: CSS;
    messageNotification: {
      container: CSS;
      text: CSS;
    };
    messageSystem: {
      container: CSS;
      dateText: CSS;
      line: CSS;
      text: CSS;
      textContainer: CSS;
    };
    typingIndicatorContainer: CSS;
  };
  spinner: CSS;
  thread: {
    newThread: CSS & {
      text: CSS;
    };
  };
  typingIndicator: {
    container: CSS;
    text: CSS & {
      color: string;
      fontSize: number;
    };
  };
};

export const defaultTheme: Theme = {
  avatar: {
    container: {},
    fallback: {},
    image: {},
    text: {},
  },
  channelListFooterLoadingIndicator: {
    container: {},
  },
  channelListHeaderErrorIndicator: {
    container: {},
    errorText: {},
  },
  channelListHeaderNetworkDownIndicator: {
    container: {},
    errorText: {},
  },
  channelPreview: {
    container: {},
    date: {},
    details: {},
    detailsTop: {},
    message: {
      color: '#767676',
      fontWeight: 'normal',
      unreadColor: '#000',
      unreadFontWeight: 'bold',
    },
    title: {},
  },
  closeButton: {
    container: {},
  },
  colors: {
    ...Colors,
  },
  iconBadge: {
    container: {},
    icon: {},
    iconInner: {},
    unreadCount: {},
  },
  iconSquare: {
    container: {},
  },
  loadingErrorIndicator: {
    container: {},
    errorText: {},
    retryText: {},
  },
  loadingIndicator: {
    container: {},
    loadingText: {},
  },
  message: {
    actions: {
      button: {
        defaultBackgroundColor: 'white',
        defaultBorderColor: 'transparent',
        primaryBackgroundColor: Colors.primary,
        primaryBorderColor: Colors.light,
      },
      buttonText: {
        defaultColor: 'black',
        primaryColor: 'white',
      },
      container: {},
    },
    actionSheet: {
      buttonContainer: {},
      buttonText: {},
      cancelButtonContainer: {},
      cancelButtonText: {},
      titleContainer: {},
      titleText: {},
    },
    avatarWrapper: {
      container: {},
      spacer: {},
    },
    card: {
      container: {},
      cover: {},
      footer: {
        description: {},
        link: {},
        logo: {},
        title: {},
      },
    },
    container: {},
    content: {
      container: {
        borderRadiusL: 16,
        borderRadiusS: 2,
      },
      containerInner: {},
      deletedContainer: {},
      deletedText: {},
      errorContainer: {
        backgroundColor: Colors.danger,
      },
      markdown: {},
      metaContainer: {},
      metaText: {},
      textContainer: {
        borderRadiusL: 16,
        borderRadiusS: 2,
        leftBorderColor: 'rgba(0, 0, 0, 0.08)',
        leftBorderWidth: 0.5,
        rightBorderColor: 'transparent',
        rightBorderWidth: 0,
      },
    },
    file: {
      container: {},
      details: {},
      icon: {},
      size: {},
      title: {},
    },
    gallery: {
      doubleSize: 240,
      galleryContainer: {},
      halfSize: 80,
      header: {
        button: {},
        container: {},
      },
      imageContainer: {},
      single: {},
      size: 120,
      width: 240,
    },
    reactionList: {
      container: {},
      reactionCount: {},
    },
    reactionPicker: {
      column: {},
      container: {},
      containerView: {},
      emoji: {},
      reactionCount: {},
      text: {},
    },
    replies: {
      container: {},
      image: {},
      messageRepliesText: {},
    },
    status: {
      checkMark: {},
      deliveredCircle: {},
      deliveredContainer: {},
      readByContainer: {},
      readByCount: {},
      sendingContainer: {},
      sendingImage: {},
      spacer: {},
    },
  },
  messageInput: {
    actionSheet: {
      buttonContainer: {},
      buttonText: {},
      titleContainer: {},
      titleText: {},
    },
    attachButton: {},
    attachButtonIcon: {},
    container: {
      conditionalPadding: 20,
    },
    editingBoxContainer: {},
    editingBoxHeader: {},
    editingBoxHeaderTitle: {},
    fileUploadPreview: {
      attachmentContainerView: {},
      attachmentView: {},
      container: {},
      dismiss: {},
      dismissImage: {},
      filenameText: {},
      itemContainer: {},
    },
    imageUploadPreview: {
      container: {},
      dismiss: {},
      dismissImage: {},
      itemContainer: {},
      upload: {},
    },
    inputBox: {},
    inputBoxContainer: {},
    sendButton: {},
    sendButtonIcon: {},
    suggestions: {
      command: {
        args: {},
        container: {},
        description: {},
        title: {},
        top: {},
      },
      container: {
        itemHeight: 50,
        maxHeight: 250,
      },
      item: {},
      mention: {
        container: {},
        name: {},
      },
      separator: {},
      title: {},
      wrapper: {},
    },
    uploadProgressIndicator: {
      container: {},
      overlay: {},
    },
  },
  messageList: {
    dateSeparator: {
      container: {},
      date: {},
      dateText: {},
      line: {},
    },
    errorNotification: {},
    errorNotificationText: {},
    eventIndicator: {
      date: {},
      memberUpdateContainer: {},
      memberUpdateText: {},
      memberUpdateTextContainer: {},
    },
    listContainer: {},
    messageNotification: {
      container: {},
      text: {},
    },
    messageSystem: {
      container: {},
      dateText: {},
      line: {},
      text: {},
      textContainer: {},
    },
    typingIndicatorContainer: {},
  },
  spinner: {},
  thread: {
    newThread: {
      text: {},
    },
  },
  typingIndicator: {
    container: {},
    text: {
      color: Colors.textGrey,
      fontSize: 14,
    },
  },
};
