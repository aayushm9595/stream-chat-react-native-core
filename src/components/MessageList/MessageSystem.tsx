import React from 'react';

import {
  isDayOrMoment,
  TDateTimeParserInput,
  useTranslationContext,
} from '../../contexts/translationContext/TranslationContext';
import { styled } from '../../styles/styledComponents';

import type { Message } from './utils/insertDates';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

const Container = styled.View`
  align-items: center;
  flex-direction: row;
  justify-content: center;
  margin-bottom: 10px;
  ${({ theme }) => theme.messageList.messageSystem.container.css}
`;

const DateText = styled.Text`
  color: rgba(0, 0, 0, 0.5);
  font-size: 10px;
  font-weight: bold;
  text-align: center;
  ${({ theme }) => theme.messageList.messageSystem.dateText.css}
`;

const Line = styled.View`
  background-color: ${({ theme }) => theme.colors.light};
  flex: 1;
  height: 0.5px;
  ${({ theme }) => theme.messageList.messageSystem.line.css}
`;

const Text = styled.Text`
  color: rgba(0, 0, 0, 0.5);
  font-size: 10px;
  font-weight: bold;
  text-align: center;
  ${({ theme }) => theme.messageList.messageSystem.text.css}
`;

const TextContainer = styled.View`
  flex: 3;
  margin-top: 10px;
  ${({ theme }) => theme.messageList.messageSystem.textContainer.css}
`;

export type MessageSystemProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = {
  /** Current [message object](https://getstream.io/chat/docs/#message_format) */
  message: Message<At, Ch, Co, Ev, Me, Re, Us>;
  /**
   * Formatter function for date object.
   *
   * @param date TDateTimeParserInput object of message
   * @returns string
   */
  formatDate?: (date: TDateTimeParserInput) => string;
};

/**
 * A component to display system message. e.g, when someone updates the channel,
 * they can attach a message with that update. That message will be available
 * in message list as (type) system message.
 */
export const MessageSystem = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  props: MessageSystemProps<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const { formatDate, message } = props;

  const { tDateTimeParser } = useTranslationContext();

  const createdAt = message.created_at as TDateTimeParserInput | undefined;
  const parsedDate = tDateTimeParser(createdAt);
  const date =
    formatDate && createdAt
      ? formatDate(createdAt)
      : parsedDate && isDayOrMoment(parsedDate)
      ? parsedDate.calendar().toUpperCase()
      : parsedDate;

  return (
    <Container testID='message-system'>
      <Line />
      <TextContainer>
        <Text>{message.text?.toUpperCase() || ''}</Text>
        <DateText>{date}</DateText>
      </TextContainer>
      <Line />
    </Container>
  );
};
