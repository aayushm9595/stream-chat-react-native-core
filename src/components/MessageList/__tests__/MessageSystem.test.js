import React from 'react';
import { ThemeProvider } from 'styled-components/native';
import { cleanup, render, waitFor } from '@testing-library/react-native';

import {
  generateMessage,
  generateStaticMessage,
} from 'mock-builders/generator/message';
import { generateStaticUser } from 'mock-builders/generator/user';

import { MessageSystem } from '../MessageSystem';

import { TranslationProvider } from '../../../contexts/translationContext/TranslationContext';
import { defaultTheme } from '../../../styles/themeConstants';
import { Streami18n } from '../../../utils/Streami18n';

afterEach(cleanup);

describe('MessageSystem', () => {
  it('should render the message system', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const message = generateMessage();
    const { queryByTestId } = render(
      <ThemeProvider theme={defaultTheme}>
        <TranslationProvider value={translators}>
          <MessageSystem message={message} />
        </TranslationProvider>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(queryByTestId('message-system')).toBeTruthy();
    });
  });

  it('should match the snapshot for message system', async () => {
    const i18nInstance = new Streami18n();
    const translators = await i18nInstance.getTranslators();
    const user = generateStaticUser(0);
    const message = generateStaticMessage('Hello World', { user });
    const { toJSON } = render(
      <ThemeProvider theme={defaultTheme}>
        <TranslationProvider value={translators}>
          <MessageSystem message={message} />
        </TranslationProvider>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
