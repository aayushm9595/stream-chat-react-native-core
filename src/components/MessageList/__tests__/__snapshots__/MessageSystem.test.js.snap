// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageSystem should match the snapshot for message system 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "center",
        "marginBottom": 10,
      },
    ]
  }
  testID="message-system"
>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#EBEBEB",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 0.5,
        },
      ]
    }
  />
  <View
    style={
      Array [
        Object {
          "flexBasis": 0,
          "flexGrow": 3,
          "flexShrink": 1,
          "marginTop": 10,
        },
      ]
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.5)",
            "fontSize": 10,
            "fontWeight": "bold",
            "textAlign": "center",
          },
        ]
      }
    >
      HELLO WORLD
    </Text>
    <Text
      style={
        Array [
          Object {
            "color": "rgba(0, 0, 0, 0.5)",
            "fontSize": 10,
            "fontWeight": "bold",
            "textAlign": "center",
          },
        ]
      }
    >
      04/27/2020
    </Text>
  </View>
  <View
    style={
      Array [
        Object {
          "backgroundColor": "#EBEBEB",
          "flexBasis": 0,
          "flexGrow": 1,
          "flexShrink": 1,
          "height": 0.5,
        },
      ]
    }
  />
</View>
`;
